src/lib/ldap/sasl.c

```diff
@@ -372,3 +354,162 @@ int fr_ldap_sasl_bind_async(fr_ldap_connection_t *c,
 	return 0;
 }
 
+
+/** Submit an async SASL LDAP auth bind
+ *
+ * @param[in] p_result	Unused.
+ * @param[in] priority	Unused.
+ * @param[in] request	being processed.
+ * @param[in] uctx	bind auth ctx.
+ * @return	unlang action.
+ */
+static unlang_action_t ldap_async_sasl_auth_bind_start(UNUSED rlm_rcode_t *p_result, UNUSED int *priority,
+						       request_t *request, void *uctx)
+{
+	fr_ldap_bind_auth_ctx_t	*bind_auth_ctx = talloc_get_type_abort(uctx, fr_ldap_bind_auth_ctx_t);
+	fr_ldap_sasl_ctx_t	*sasl_ctx = bind_auth_ctx->sasl_ctx;
+	int			ret;
+
+	RDEBUG2("%s SASL bind auth operation as %s", sasl_ctx->rmech ? "Continuing" : "Starting", sasl_ctx->dn);
+
+	ret = ldap_sasl_interactive_bind(sasl_ctx->c->handle, sasl_ctx->dn, sasl_ctx->mechs,
+					 NULL, NULL, LDAP_SASL_AUTOMATIC,
+					 _sasl_interact, sasl_ctx, sasl_ctx->result,
+					 &sasl_ctx->rmech, &bind_auth_ctx->msgid);
+
+	switch (ret) {
+	case LDAP_SUCCESS:
+		return UNLANG_ACTION_CALCULATE_RESULT;
+
+	case LDAP_SASL_BIND_IN_PROGRESS:
+		fr_rb_insert(bind_auth_ctx->thread->binds, bind_auth_ctx);
+		return UNLANG_ACTION_YIELD;
+
+	default:
+		return UNLANG_ACTION_FAIL;
+	}
+}
+
+/** Signal an outstanding SASL LDAP bind to cancel
+ *
+ * @param[in] request	being processed. Unused.
+ * @param[in] action	Signal to handle.
+ * @param[in] uctx	bind auth ctx.
+ */
+static void ldap_async_sasl_auth_bind_cancel(UNUSED request_t *request, fr_state_signal_t action, void *uctx)
+{
+	fr_ldap_bind_auth_ctx_t *bind_auth_ctx = talloc_get_type_abort(uctx, fr_ldap_bind_auth_ctx_t);
+
+	if (action != FR_SIGNAL_CANCEL) return;
+
+	ldap_abandon_ext(bind_auth_ctx->sasl_ctx->c->handle, bind_auth_ctx->msgid, NULL, NULL);
+	fr_rb_remove(bind_auth_ctx->thread->binds, bind_auth_ctx);
+
+	talloc_free(bind_auth_ctx);
+}
+
+/** Handle the return code from parsed LDAP results to set the module rcode
+ *
+ * @param[out] p_result	Where to write return code.
+ * @param[in] priority	Unused.
+ * @param[in] request	being processed.
+ * @param[in] uctx	bind auth ctx.
+ * @return	unlang action.
+ */
+static unlang_action_t ldap_async_sasl_auth_bind_results(rlm_rcode_t *p_result, UNUSED int *priority, request_t *request, void *uctx)
+{
+	fr_ldap_bind_auth_ctx_t	*bind_auth_ctx = talloc_get_type_abort(uctx, fr_ldap_bind_auth_ctx_t);
+	fr_ldap_sasl_ctx_t	*sasl_ctx = bind_auth_ctx->sasl_ctx;
+
+	switch (bind_auth_ctx->ret) {
+	case LDAP_PROC_SUCCESS:
+		RDEBUG2("Bind as user \"%s\" was successful", sasl_ctx->identity);
+		break;
+
+	case LDAP_PROC_NOT_PERMITTED:
+		RDEBUG2("Bind as user \"%s\" not permitted", sasl_ctx->identity);
+		break;
+
+	case LDAP_PROC_REJECT:
+		RDEBUG2("Bind as user \"%s\" rejected", sasl_ctx->identity);
+		break;
+
+	case LDAP_PROC_CONTINUE:
+		return unlang_function_push(request, ldap_async_sasl_auth_bind_start, ldap_async_sasl_auth_bind_results,
```

Comment:
Maybe we are missing to release the pointers here:

```
	ldap_msgfree(sasl_ctx->result);
	talloc_free(bind_auth_ctx);
```

Reply:
No, the bind_auth_ctx gets reused in the call to unlang_function_push, and is parented from a stack frame so can never leak.  You may need to release `sasl_ctx->result` though @ndptech?

Reply:
No - at that point `sasl_ctx->result` points at the last result, which is passed back to `ldap_sasl_interactive_bind`.
When you have a multi-round SASL auth, the result of the previous round is freed in `_ldap_bind_auth_io_read()` in `rlm_ldap.c` before the next result is retrieved.

This follows the same model as multi round SASL binds as the admin user.

Reply:
OK

Reply:
You'll still leak `sasl_ctx->result` if the request is cancelled between yielding and getting a response back from LDAP as there's no cleanup code in the signal callback.  It might be worth adding destructors to any ctxs that hold an `ldap_result` to make sure that if the pointer isn't NULL, that it's always passed to `ldap_msgfree` if the ctx gets freed.

Reply:
Would probably also leak on I/O timeout?